runtime/src/iree/builtins/ukernel/arch/arm_64/mmt4d_arm_64_dotprod.c

```diff
@@ -75,3 +75,109 @@ IREE_UK_MMT4D_TILE_FUNC_IMPL_FOR_M0_1_2_4_8(
     iree_uk_mmt4d_tile_s8s8s32_2x8x4_arm_64_dotprod,
     iree_uk_mmt4d_tile_s8s8s32_4x8x4_arm_64_dotprod,
     iree_uk_mmt4d_tile_s8s8s32_8x8x4_arm_64_dotprod)
+
+static inline void iree_uk_mmt4d_tile_s8s4s32_1x8x8_to_4x8x8_arm_64_dotprod(
+    void* IREE_UK_RESTRICT out_tile, const void* IREE_UK_RESTRICT lhs_panel,
+    const void* IREE_UK_RESTRICT rhs_panel,
+    const iree_uk_mmt4d_params_t* params, int M0) {
+  IREE_UK_ASSERT(M0 >= 1 && M0 <= 8 && iree_uk_is_po2_u32(M0));
+  const iree_uk_int8_t* IREE_UK_RESTRICT lhs_ptr = lhs_panel;
+  const iree_uk_int8_t* IREE_UK_RESTRICT rhs_ptr = rhs_panel;
+  iree_uk_int32_t* IREE_UK_RESTRICT out_ptr = out_tile;
+
+  const int8x16_t vmask = vmovq_n_s8(0xF0);
+  const int8x8_t vzero = vmov_n_s8(0);
+
+  int32x4_t acc[8];
+  for (int i = 0; i < 8; i++) {
+    // We start with zero accumulators and add the value of *out_ptr later.
+    // This is required for the int4 left shift described later.
+    acc[i] = vdupq_n_s32(0);
+  }
+
+  for (int k = 0; k < params->K; ++k) {
+    int8x16_t r1 = vld1q_s8(rhs_ptr);
+    rhs_ptr += 16;
+    int8x16_t r2 = vld1q_s8(rhs_ptr);
+    rhs_ptr += 16;
+
+    // We unpack int4s into individual int8s. To preserve signedness,
+    // int4s are moved to the upper 4-bits of each byte. This has the effect of
+    // multiplying each int4 by 2^4 = 16. To compensate, we divide the
+    // accumulator values by 16 before storing to memory.
+    // This int4 conversion trick is borrowed from the `qd8-f32-qc4w-gemm*`
+    // kernels in https://github.com/google/XNNPACK.
+    int8x16_t rhs[4];
+    rhs[0] = vshlq_n_s8(r1, 4);
+    rhs[1] = vshlq_n_s8(r2, 4);
+    rhs[2] = vandq_s8(r1, vmask);
+    rhs[3] = vandq_s8(r2, vmask);
+
+    if (M0 == 4) {
```

Comment:
instead of if M== 4 etc within the K loop, consider
a) k loops inside each if
b) completely different kernels and/or
c) handle max size (M=8) and somehow ignore the unused M values.

Reply:
> a) k loops inside each if

If I did k-loops inside of each, I would need to repeat loading of the RHS for each loop. At the moment, this code is shared across all M0 values.

> b) completely different kernels and/or

We are limited to the data-tiling layout here and I can't immediately think of a better kernel for each M0 size.

> c) handle max size (M=8) and somehow ignore the unused M values.

I just added an M0=8 and it's looking verbose. I'll try and consolidate the code across the M0 values.

Reply:
Refactored and reduced LOC. Performance is about the same after refactoring.